Report unused imports

Akirathan opened the pull request:
Fixes #12825

### Pull Request Description

<!--
- Please describe the nature of your PR here, as well as the motivation for it.
- If it fixes an open issue, please mention that issue number here.
-->

### Important Notes

- Report unused imports

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

Commit:
Extract Mock compiler ctx classes from nested classes to separate package

Commit:
CouldNotCreateDirectory exception has cause

Commit:
Implement VirtualFileSystem based on Apache Commons VFS2

Commit:
Implement WithMockCompilerContext TestRule

Commit:
fmt

Commit:
Print compiler output in case of failure

Commit:
fmt

Commit:
runtime-compiler/test runs with logging-service-logback/Test logger

Commit:
Fix file reading in MockPackageRepository

Commit:
typo

Commit:
Can resolve imports

Akirathan commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/WithMockCompilerContext.java:
```diff
@@ -0,0 +1,160 @@
+ * <p>I recommend to use it as a {@link org.junit.Rule}, instead of {@link org.junit.ClassRule} to
+ * ensure that the whole context is reset before each test.
+ */
+public final class WithMockCompilerContext implements TestRule {
```
This TestRule is the entrypoint for the tests. Compiler tests are expected to use all the mocked classes only via this class (note that all other classes are package private).

Akirathan replied:
The API is still work in progress, but I believe the main idea is solid. Moreover, all the tests pass locally.

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/UnusedImportsTest.java:
```diff
@@ -0,0 +1,80 @@
+import org.junit.Test;
+
+public class UnusedImportsTest {
+  @Rule public final WithMockCompilerContext compilerCtx = WithMockCompilerContext.createDefault();
```
```suggestion
  @Rule private final WithMockCompilerContext compilerCtx = WithMockCompilerContext.createDefault();
```

Akirathan replied:
Field annotated with `@Rule` must be public:
```
[error] Test org.enso.compiler.test.UnusedImportsTest.initializationError failed: org.junit.runners.model.InvalidTestClassError: Invalid test class 'org.enso.compiler.test.UnusedImportsTest':
```

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/UnusedImportsTest.java:
```diff
@@ -0,0 +1,80 @@
+
+  private CompilerContext.Module getMainModule() {
+    return compilerCtx.getLoadedModules().stream()
+        .filter(mod -> mod.getName().item().equals("Main"))
```
Isn't this going to be pretty common operation? Maybe:
```suggestion
        compilerCtx.findModule("Main")
```

Akirathan replied:
Good point. Implemented in bb0ec19af40aca5abcf480ecb163db7d8148ca7f

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/MockModule.java:
```diff
@@ -0,0 +1,125 @@
+import org.enso.pkg.Package;
+import org.enso.pkg.QualifiedName;
+
+final class MockModule extends CompilerContext.Module {
```
When I was considering the refactoring, I wanted to put `MocModule` as an inner class of `MockCompilerContext`.... but this is fine as well.

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/MockPackageRepository.java:
```diff
@@ -0,0 +1,289 @@
+ * {@link PackageRepository} implementation with {@link org.apache.commons.vfs2.FileObject} as its
+ * type member.
+ */
+final class MockPackageRepository implements PackageRepository {
```
Should this repository "see" `Standard.Base`? Or should it be (stay) empty?

Akirathan replied:
As long as this repository stays in tests, I would prefer it not having access to `Standard.Base`. If, in the future, we add some tests that explicitly require `Standard.Base` (I doubt that we need std libs for testing compiler), we could add this functionality - should not be difficult with the virtual file system.

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/SourceModule.java:
```diff
@@ -0,0 +1,5 @@
+
+import org.enso.pkg.QualifiedName;
+
+public record SourceModule(QualifiedName name, String content) {}
```
Shouldn't one use `MockModule` instead of `SourceModule`? E.g. have a method `CompilerContext.Module WithMockCompilerContext.createModule(QualifiedName, String)`?

What's a benefit of having one more layer of indirection?

Akirathan replied:
Removed in 57e479ee5cb96251ac9d41faeed49d07c5d59843

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/VirtualFileSystem.java:
```diff
@@ -0,0 +1,218 @@
+ * {@link FileSystem} implementation for Apache Commons VFS2. So far, only RAM file system is
+ * supported.
+ */
+final class VirtualFileSystem implements FileSystem<FileObject> {
```
- using 3rd party API class in the signature is diverging us from my dreamed of direction
- using `java.nio.Path` instead in the API signature would avoid such 3rd party API dependency
- using [jimfs](https://github.com/google/jimfs) as the in memory implementation would be 1:1 rewrite

Akirathan replied:
Replaced Apache Commons VFS2 with com.google.jimfs in ae8e0f985b03dcdafc308cf24884cee120c378f6. Now, we have `VirtualFileSystem` that implements `org.enso.filesystem.FileSystem<java.nio.file.Path>`. It is also much easier to use.

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/VirtualFileSystemTest.java:
```diff
@@ -0,0 +1,162 @@
+import org.junit.rules.TemporaryFolder;
+
+public class VirtualFileSystemTest {
+  @ClassRule public static final TemporaryFolder TMP_DIR = new TemporaryFolder();
```
Why do we need temporary folder? Is the variable even used?

Akirathan replied:
It was a leftover. Removed in 116cbcc6994ae5260a1e91d174fc50283bc81227

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/WithMockCompilerContext.java:
```diff
@@ -0,0 +1,160 @@
+import scala.jdk.javaapi.CollectionConverters;
+
+/**
+ * A {@link TestRule} to use with the {@link MockCompilerContext}.
```
`MockCompilerContext` is an implementation class, not visible in the API. It makes little sense to reference it from API documentation.

Akirathan replied:
Docs improved in caa8e756ba79bad4d07dca2f90b4e9c877e6ef4a

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/WithMockCompilerContext.java:
```diff
@@ -0,0 +1,160 @@
+    return compiler;
+  }
+
+  public Package<FileObject> createPackage(LibraryName pkgName, SourceModule module) {
```
What is the purpose of `SourceModule`?

Akirathan replied:
None. Removed in 57e479ee5cb96251ac9d41faeed49d07c5d59843

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/WithMockCompilerContext.java:
```diff
@@ -0,0 +1,160 @@
+   * @param modules
+   * @return Created package, or null if an error occurred.
+   */
+  public Package<FileObject> createPackage(LibraryName pkgName, Set<SourceModule> modules) {
```
Why not have just a `Module` factory method?
```suggestion
  public Module createModule(QualifiedName name, String context) {
```
which would put the module to the right `Package` based on the provided FQN `name`?

Akirathan replied:
Refactored in 57e479ee5cb96251ac9d41faeed49d07c5d59843

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/WithMockCompilerContext.java:
```diff
@@ -0,0 +1,160 @@
+    return new Builder();
+  }
+
+  public MockPackageRepository getRepo() {
```
`MockPackageRepository` isn't (intentionally) a public class. Return just `PackageRepository` as the API class here.
```suggestion
  public PackageRepository<Path> getPackageRepository() {
```
probably better to use full getter name than shorten it up arbitrarily.

Akirathan replied:
In 57e479ee5cb96251ac9d41faeed49d07c5d59843, `getRepo` was removed altogether. If we will need it in the future, we can implement your suggestion.

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/mock/WithMockCompilerContext.java:
```diff
@@ -0,0 +1,160 @@
+    public WithMockCompilerContext build() {
+      var out = new ByteArrayOutputStream();
+      var compilerCfg =
+          new CompilerConfig(
```
I was also looking at `CompilerConfig`. It's use from Java isn't easy. You are mitigating that problem by hiding the config into new builder in tests.

Why not rewrite `CompilerConfig` into Java and turn it into builder pattern itself? Then you'd have:
```
public static WithMockCompilerContext create(CompilerConfig) {
```

Akirathan replied:
`CompilerConfig` refactored to Java in 341ca62ec2ef61d8aebeaed18f69b436351d4d90

JaroslavTulach reviewed (commented):
I would prefer usage of `java.nio.Path` in the API instead of 3rd party library class.

Commit:
Implement findModule method

Commit:
Remove unnecessary toString method calls

Commit:
Move createPackage functionality into single createModuleMethod.

Remove unnecessary SourceModule record.

Commit:
Remove unnecessary ClassRule

Commit:
Improve docs

Commit:
Refactor CompilerConfig to Java

Commit:
Use com.google.jimfs instead of org.apache.commons.vfs2.

Commit:
Fix usage example of IGV for compiler IR

Commit:
Implement MockModule.getUri

Commit:
Add implementation of UnusedImports pass

Commit:
Add simple test

Commit:
Move runtime-compiler tests to a separate project.

This will be necessary if we want to dump IGV graphs during compilation. Otherwise, there is a cyclic dependency in sbt.

Commit:
runtime-compiler/test configuration is the same as on develop

Commit:
runtime-compiler-tests uses truffleRunOptionsSettings

Commit:
Can dump IR with IGV in runtime-compiler-tests

Commit:
Fix compilation of PassesTest

Commit:
Dummy implementation for MockModule.findLine

Commit:
fmt

Akirathan commented:
Waiting until https://github.com/enso-org/enso/pull/12992 is merged. Will rebase then.

Akirathan commented:
This PR so far contains only changes that were introduced in #12992. Closing this now and will replace this by a fresh new PR.
